notify/rocketchat/rocketchat.go

```diff
@@ -0,0 +1,266 @@
+		if err != nil {
+			return "", fmt.Errorf("could not read %s: %w", c.TokenIDFile, err)
+		}
+		return strings.TrimSpace(string(content)), nil
+	}
+	return string(*c.TokenID), nil
+}
+
+func getToken(c *config.RocketchatConfig) (string, error) {
+	if len(c.TokenFile) > 0 {
+		content, err := os.ReadFile(c.TokenFile)
+		if err != nil {
+			return "", fmt.Errorf("could not read %s: %w", c.TokenFile, err)
+		}
+		return strings.TrimSpace(string(content)), nil
+	}
+	return string(*c.Token), nil
+}
+
+// Notify implements the Notifier interface.
+func (n *Notifier) Notify(ctx context.Context, as ...*types.Alert) (bool, error) {
+	var err error
+
+	data := notify.GetTemplateData(ctx, n.tmpl, as, n.logger)
+	tmplText := notify.TmplText(n.tmpl, data, &err)
+	if err != nil {
+		return false, err
+	}
+	titlex := tmplText(n.conf.Title)
+	if err != nil {
+		return false, err
+	}
+
+	title, truncated := notify.TruncateInRunes(titlex, maxTitleLenRunes)
+	if truncated {
+		key, err := notify.ExtractGroupKey(ctx)
+		if err != nil {
+			return false, err
+		}
+		level.Warn(n.logger).Log("msg", "Truncated title", "key", key, "max_runes", maxTitleLenRunes)
+	}
+	att := &Attachment{
+		Title:     title,
+		TitleLink: tmplText(n.conf.TitleLink),
+		Text:      tmplText(n.conf.Text),
+		ImageURL:  tmplText(n.conf.ImageURL),
+		ThumbURL:  tmplText(n.conf.ThumbURL),
+		Color:     tmplText(n.conf.Color),
+	}
+	numFields := len(n.conf.Fields)
+	if numFields > 0 {
+		fields := make([]config.RocketchaAttachmentField, numFields)
+		for index, field := range n.conf.Fields {
+			// Check if short was defined for the field otherwise fallback to the global setting
+			var short bool
+			if field.Short != nil {
+				short = *field.Short
+			} else {
+				short = n.conf.ShortFields
+			}
+
+			// Rebuild the field by executing any templates and setting the new value for short
+			fields[index] = config.RocketchaAttachmentField{
+				Title: tmplText(field.Title),
+				Value: tmplText(field.Value),
+				Short: &short,
+			}
+		}
+		att.Fields = fields
+	}
+	numActions := len(n.conf.Actions)
+	if numActions > 0 {
+		actions := make([]config.RocketchatAttachmentAction, numActions)
+		for index, action := range n.conf.Actions {
+			rocketchatAction := config.RocketchatAttachmentAction{
+				Type: "button", // Only button type is supported
+				Text: tmplText(action.Text),
+				URL:  tmplText(action.URL),
+				Msg:  tmplText(action.Msg),
+			}
+
+			actions[index] = rocketchatAction
+		}
+		att.Actions = actions
+	}
+
+	body := &PostMessage{
+		Channel:     n.conf.Channel,
+		Emoji:       tmplText(n.conf.Emoji),
+		Avatar:      tmplText(n.conf.IconURL),
+		Attachments: []Attachment{*att},
+	}
+	if err != nil {
+		return false, err
+	}
+
+	var buf bytes.Buffer
+	if err := json.NewEncoder(&buf).Encode(body); err != nil {
+		return false, err
+	}
+	req, err := http.NewRequest("POST", fmt.Sprintf("%s/%s", n.conf.APIURL.String(), "api/v1/chat.postMessage"), &buf)
```

Comment:
You should use the `postJSONFunc` like in https://github.com/prometheus/alertmanager/blob/main/notify/slack/slack.go#L208. I think we can update it to accept headers or even `*http.Request`?

Reply:
Can you please point me at how to do this correctly? `notify.PostJSON` is used by some integrations but not all. Do you suggest I should change the function signature of `notify.PostJSON`, and  thus change `notify.post` and `notify.request`?

Reply:
I was thinking of adding a `headers` argument which would be of the type `http.Headers`. Let me check to see how big of a change it is, as I might do it in a separate PR which you can then rebase.

Reply:
I was tinkering with exactly that yesterday evening and to me it looked like a very substantial change, thus the question.

Reply:
This is what I thought https://github.com/prometheus/alertmanager/pull/3776.

Reply:
jep looks pretty much how I expected/encountered on tinkering

Reply:
should be addressed in a acceptable manner now